------------------------------------------------------------
revno: 1782
committer: David Kalnischkies <kalnischkies@gmail.com>
branch nick: apt
timestamp: Thu 2009-05-07 23:50:11 +0200
message:
  [apt-get] add possibility to reorder the warning section (Warning-Order)
  * apply on all operations installing (or removing) packages
  * two translatable strings (malformed errormessages) and a bit of doc
  * Suggest and Recommend is not included in this reordering (for now)
diff:
=== modified file 'cmdline/apt-get.cc'
--- cmdline/apt-get.cc	2009-04-14 12:17:40 +0000
+++ cmdline/apt-get.cc	2009-05-07 21:50:11 +0000
@@ -753,18 +753,51 @@
    bool Essential = false;
    
    // Show all the various warning indicators
-   ShowDel(c1out,Cache);
-   ShowNew(c1out,Cache);
-   if (ShwKept == true)
-      ShowKept(c1out,Cache);
-   Fail |= !ShowHold(c1out,Cache);
-   if (_config->FindB("APT::Get::Show-Upgraded",true) == true)
-      ShowUpgraded(c1out,Cache);
-   Fail |= !ShowDowngraded(c1out,Cache);
-   if (_config->FindB("APT::Get::Download-Only",false) == false)
-        Essential = !ShowEssential(c1out,Cache);
-   Fail |= Essential;
-   Stats(c1out,Cache);
+   string WarningOrder = _config->Find("APT::Get::Warning-Order",
+      "Remove,New,Kept,Hold,Upgraded,Downgraded,Essential,Stats")+",";
+   short printedSection = 0;
+   size_t NextComma = string::npos;
+   while((NextComma = WarningOrder.find(','))!=string::npos)
+   {
+      string DisplaySection = WarningOrder.substr(0,NextComma);
+      WarningOrder = WarningOrder.substr(NextComma+1);
+      if (DisplaySection == "Remove")
+         ShowDel(c1out,Cache);
+      else if (DisplaySection == "New")
+         ShowNew(c1out,Cache);
+      else if (DisplaySection == "Kept")
+      {
+         if (ShwKept == true)
+            ShowKept(c1out,Cache);
+      }
+      else if (DisplaySection == "Hold")
+         Fail |= !ShowHold(c1out,Cache);
+      else if (DisplaySection == "Upgraded")
+      {
+         if (_config->FindB("APT::Get::Show-Upgraded",true) == true)
+            ShowUpgraded(c1out,Cache);
+      }
+      else if (DisplaySection == "Downgraded")
+         Fail |= !ShowDowngraded(c1out,Cache);
+      else if (DisplaySection == "Essential")
+      {
+         if (_config->FindB("APT::Get::Download-Only",false) == false)
+            Essential = !ShowEssential(c1out,Cache);
+         Fail |= Essential;
+      }
+      else if (DisplaySection == "Stats")
+         Stats(c1out,Cache);
+      else
+         return _error->Error(_("Unknown section \"%s\" specified in APT::Get::Warning-Order!"),
+            DisplaySection.c_str());
+
+      printedSection++;
+   }
+
+   if (printedSection < 8)
+      return _error->Error(
+         _("APT::Get::Warning-Order is malformed! Only %d sections specified; required: %d sections!"),
+         printedSection,8);
 
    // Sanity check
    if (Cache->BrokenCount() != 0)

=== modified file 'doc/apt-get.8.xml'
--- doc/apt-get.8.xml	2009-05-07 10:04:21 +0000
+++ doc/apt-get.8.xml	2009-05-07 21:50:11 +0000
@@ -515,10 +515,18 @@
      This is useful for tools like pbuilder.
      Configuration Item: <literal>APT::Get::AllowUnauthenticated</literal>.</para></listitem>
      </varlistentry>
-     
 
      &apt-commonoptions;
-     
+
+     <varlistentry><term><literal>APT::Get::Warning-Order</literal></term>
+     <listitem><para>Used to specify the order of the sections warning a user
+     about (possible) new, kept, removed and such packages. Default setting is
+     "<literal>Remove,New,Kept,Hold,Upgraded,Downgraded,Essential,Stats</literal>".
+     Note that this option should be used to change the order, it is not suitable to
+     hide sections! &apt-get; will refuse to work if this setting is malformed.
+     </para></listitem>
+     </varlistentry>
+
    </variablelist>
  </refsect1>
